File: standalone-container/src/main/scala/com/yahoo/application/container/impl/ClassLoaderOsgiFramework.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.application.container.impl

import java.io.InputStream
import java.net.{URL, URLClassLoader}
import java.util
import java.util.concurrent.atomic.AtomicInteger
import java.util.jar.JarFile
import java.util.{Collections, Dictionary, Hashtable}

import com.yahoo.container.standalone.StandaloneContainerApplication
import com.yahoo.jdisc.application.{OsgiFramework, OsgiHeader}
import org.osgi.framework._
import org.osgi.framework.wiring._
import org.osgi.resource.{Capability, Requirement, Wire}

import scala.collection.JavaConverters._
import scala.collection.mutable.ArrayBuffer

/**
 * A (mock) OSGI implementation which loads classes from the system classpath
 * 
 * @author tonytv
 */
final class ClassLoaderOsgiFramework extends OsgiFramework {
  private val bundleLocations = new ArrayBuffer[URL]
  private val bundleList = ArrayBuffer[Bundle](SystemBundleImpl)
  private var classLoader: ClassLoader = null

  private val nextBundleId = new AtomicInteger(1)

  override def installBundle(bundleLocation: String) = {
    if (bundleLocation != "") {
      val url = new URL(bundleLocation)
      bundleLocations += url
      bundleList += new JarBundleImpl(url)
    }

    bundles()
  }

  def getClassLoader = {
    if (bundleLocations.isEmpty) {
      getClass.getClassLoader
    } else {
      if(classLoader == null)
        classLoader = new URLClassLoader(bundleLocations.toArray, getClass.getClassLoader)

      classLoader
    }
  }

  override def startBundles(bundles: util.List[Bundle], privileged: Boolean) {}

  override def refreshPackages() {}

  override def bundleContext():BundleContext = BundleContextImpl

  override def bundles() = bundleList.asJava

  override def start() {}

  override def stop() {}

  private abstract class BundleImpl extends Bundle {
    override def getState = Bundle.ACTIVE

    override def start(options: Int) {}
    override def start() {}
    override def stop(options: Int) {}
    override def stop() {}
    override def update(input: InputStream) {}
    override def update() {}
    override def uninstall() {}

    override def getHeaders(locale: String) = getHeaders

    override def getSymbolicName = ClassLoaderOsgiFramework.this.getClass.getName
    override def getLocation = getSymbolicName

    override def getRegisteredServices = Array[ServiceReference[_]]()
    override def getServicesInUse = getRegisteredServices

    override def hasPermission(permission: Any) = true

    override def getResource(name: String) = getClassLoader.getResource(name)
    override def loadClass(name: String) = getClassLoader.loadClass(name)
    override def getResources(name: String) = getClassLoader.getResources(name)

    override def getEntryPaths(path: String) = throw new UnsupportedOperationException
    override def getEntry(path: String) = throw new UnsupportedOperationException
    override def findEntries(path: String, filePattern: String, recurse: Boolean) = throw new UnsupportedOperationException

    override def getLastModified = 1L

    override def getBundleContext = throw new UnsupportedOperationException
    override def getSignerCertificates(signersType: Int) = Collections.emptyMap()

    override def adapt[A](`type`: Class[A]): A = {
      if (`type` == classOf[BundleRevision]) BundleRevisionImpl.asInstanceOf[A]
      else if (`type` == classOf[BundleWiring]) BundleWiringImpl.asInstanceOf[A]
      else null.asInstanceOf[A]
    }

    override def getDataFile(filename: String) = null
    override def compareTo(o: Bundle) = getBundleId compareTo o.getBundleId
  }

  private object BundleRevisionImpl extends BundleRevision {
    override def getSymbolicName: String = this.getClass.getName
    override def getDeclaredRequirements(p1: String): util.List[BundleRequirement] = throw new UnsupportedOperationException
    override def getVersion: Version = Version.emptyVersion
    override def getWiring: BundleWiring = BundleWiringImpl
    override def getDeclaredCapabilities(p1: String): util.List[BundleCapability] = throw new UnsupportedOperationException
    override def getTypes: Int = 0
    override def getBundle: Bundle = throw new UnsupportedOperationException
    override def getCapabilities(p1: String): util.List[Capability] = throw new UnsupportedOperationException
    override def getRequirements(p1: String): util.List[Requirement] = throw new UnsupportedOperationException
  }

  private object BundleWiringImpl extends BundleWiring {
    override def findEntries(p1: String, p2: String, p3: Int): util.List[URL] = ???
    override def getRequiredResourceWires(p1: String): util.List[Wire] = ???
    override def getResourceCapabilities(p1: String): util.List[Capability] = ???
    override def isCurrent: Boolean = ???
    override def getRequiredWires(p1: String): util.List[BundleWire] = ???
    override def getCapabilities(p1: String): util.List[BundleCapability] = ???
    override def getProvidedResourceWires(p1: String): util.List[Wire] = ???
    override def getProvidedWires(p1: String): util.List[BundleWire] = ???
    override def getRevision: BundleRevision = ???
    override def getResourceRequirements(p1: String): util.List[Requirement] = ???
    override def isInUse: Boolean = ???
    override def listResources(p1: String, p2: String, p3: Int): util.Collection[String] = ???
    override def getClassLoader: ClassLoader = ClassLoaderOsgiFramework.this.getClassLoader
    override def getRequirements(p1: String): util.List[BundleRequirement] = ???
    override def getResource: BundleRevision = ???
    override def getBundle: Bundle = ???
  }

  private object SystemBundleImpl extends BundleImpl {
    override val getBundleId = 0L
    override def getVersion = Version.emptyVersion
    override def getHeaders: Dictionary[String, String] = new Hashtable[String, String](Map(OsgiHeader.APPLICATION -> classOf[StandaloneContainerApplication].getName).asJava)
  }


  private class JarBundleImpl(location: URL) extends BundleImpl {
    override val getBundleId = nextBundleId.getAndIncrement.asInstanceOf[Long]

    private val headers = retrieveHeaders(location)

    override def getHeaders: Dictionary[String, String] = headers
    override val getSymbolicName = headers.get("Bundle-SymbolicName")
    override val getVersion = Version.parseVersion(headers.get("Bundle-Version"))


    private def retrieveHeaders(location: URL) = {
      val jarFile = new JarFile(location.getFile)
      try {
        val attributes = jarFile.getManifest.getMainAttributes
        new Hashtable[String, String](attributes.entrySet().asScala.map( entry => entry.getKey.toString -> entry.getValue.toString).toMap.asJava)
      } finally {
        jarFile.close()
      }
    }
  }

  private object BundleContextImpl extends BundleContext {
    private val bundleImpl = SystemBundleImpl

    override def getProperty(key: String) = null
    override def getBundle = bundleImpl
    override def installBundle(location: String, input: InputStream) = throw new UnsupportedOperationException
    override def installBundle(location: String) = throw new UnsupportedOperationException

    override def getBundle(id: Long) = bundleImpl
    override def getBundles = Array(bundleImpl)
    override def getBundle(location: String) = bundleImpl

    override def addServiceListener(listener: ServiceListener, filter: String) {}
    override def addServiceListener(listener: ServiceListener) {}
    override def removeServiceListener(listener: ServiceListener) {}
    override def addBundleListener(listener: BundleListener) {}
    override def removeBundleListener(listener: BundleListener) {}

    override def addFrameworkListener(listener: FrameworkListener) {}
    override def removeFrameworkListener(listener: FrameworkListener) {}

    override def registerService(clazzes: Array[String], service: Any, properties: Dictionary[String, _]) = throw new UnsupportedOperationException
    override def registerService(clazz: String, service: Any, properties: Dictionary[String, _]) = null
    override def registerService[S](clazz: Class[S], service: S, properties: Dictionary[String, _]) = throw new UnsupportedOperationException
    override def getServiceReferences(clazz: String, filter: String) = throw new UnsupportedOperationException
    override def getAllServiceReferences(clazz: String, filter: String) = throw new UnsupportedOperationException
    override def getServiceReference(clazz: String) = throw new UnsupportedOperationException
    override def getServiceReference[S](clazz: Class[S]) = throw new UnsupportedOperationException
    override def getServiceReferences[S](clazz: Class[S], filter: String) = Collections.emptyList()
    override def getService[S](reference: ServiceReference[S]) = throw new UnsupportedOperationException
    override def ungetService(reference: ServiceReference[_]) = throw new UnsupportedOperationException
    override def getDataFile(filename: String) = throw new UnsupportedOperationException
    override def createFilter(filter: String) = throw new UnsupportedOperationException

    override def registerService[S](aClass: Class[S], serviceFactory: ServiceFactory[S], dictionary: Dictionary[String, _]): ServiceRegistration[S] = throw new UnsupportedOperationException
    override def getServiceObjects[S](serviceReference: ServiceReference[S]): ServiceObjects[S] = throw new UnsupportedOperationException
  }

}

File: standalone-container/src/main/java/com/yahoo/application/container/impl/ClassLoaderOsgiFramework.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.application.container.impl;

import com.google.common.collect.Lists;
import com.yahoo.container.standalone.StandaloneContainerApplication;
import com.yahoo.jdisc.application.OsgiFramework;
import com.yahoo.jdisc.application.OsgiHeader;
import org.osgi.framework.Bundle;
import org.osgi.framework.BundleContext;
import org.osgi.framework.BundleListener;
import org.osgi.framework.Filter;
import org.osgi.framework.FrameworkListener;
import org.osgi.framework.ServiceFactory;
import org.osgi.framework.ServiceListener;
import org.osgi.framework.ServiceObjects;
import org.osgi.framework.ServiceReference;
import org.osgi.framework.ServiceRegistration;
import org.osgi.framework.Version;
import org.osgi.framework.wiring.BundleCapability;
import org.osgi.framework.wiring.BundleRequirement;
import org.osgi.framework.wiring.BundleRevision;
import org.osgi.framework.wiring.BundleWire;
import org.osgi.framework.wiring.BundleWiring;
import org.osgi.resource.Capability;
import org.osgi.resource.Requirement;
import org.osgi.resource.Wire;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.net.URLClassLoader;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Dictionary;
import java.util.Enumeration;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.jar.Attributes;
import java.util.jar.JarFile;

/**
 * A (mock) OSGI implementation which loads classes from the system classpath
 *
 * @author Tony Vaagenes
 * @author ollivir
 */
public final class ClassLoaderOsgiFramework implements OsgiFramework {
    private BundleContextImpl bundleContextImpl = new BundleContextImpl();
    private SystemBundleImpl systemBundleImpl = new SystemBundleImpl();
    private BundleWiringImpl bundleWiringImpl = new BundleWiringImpl();

    private List<URL> bundleLocations = new ArrayList<>();
    private List<Bundle> bundleList = Lists.newArrayList(systemBundleImpl);
    private ClassLoader classLoader = null;

    private AtomicInteger nextBundleId = new AtomicInteger(1);

    @Override
    public List<Bundle> installBundle(String bundleLocation) {
        if (bundleLocation != null && bundleLocation.isEmpty() == false) {
            try {
                URL url = new URL(bundleLocation);
                bundleLocations.add(url);
                bundleList.add(new JarBundleImpl(url));
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        return bundles();
    }

    private ClassLoader getClassLoader() {
        if (bundleLocations.isEmpty()) {
            return getClass().getClassLoader();
        } else {
            if (classLoader == null) {
                classLoader = new URLClassLoader(bundleLocations.toArray(new URL[0]), getClass().getClassLoader());
            }
            return classLoader;
        }
    }

    @Override
    public void startBundles(List<Bundle> bundles, boolean privileged) {
    }

    @Override
    public void refreshPackages() {
    }

    @Override
    public BundleContext bundleContext() {
        return bundleContextImpl;
    }

    @Override
    public List<Bundle> bundles() {
        return bundleList;
    }

    @Override
    public void start() {
    }

    @Override
    public void stop() {
    }

    private abstract class BundleImpl implements Bundle {
        @Override
        public int getState() {
            return Bundle.ACTIVE;
        }

        @Override
        public void start(int options) {
        }

        @Override
        public void start() {
        }

        @Override
        public void stop(int options) {
        }

        @Override
        public void stop() {
        }

        @Override
        public void update(InputStream input) {
        }

        @Override
        public void update() {
        }

        @Override
        public void uninstall() {
        }

        @Override
        public Dictionary<String, String> getHeaders(String locale) {
            return getHeaders();
        }

        @Override
        public String getSymbolicName() {
            return ClassLoaderOsgiFramework.this.getClass().getName();
        }

        @Override
        public String getLocation() {
            return getSymbolicName();
        }

        @Override
        public ServiceReference<?>[] getRegisteredServices() {
            return new ServiceReference<?>[0];
        }

        @Override
        public ServiceReference<?>[] getServicesInUse() {
            return getRegisteredServices();
        }

        @Override
        public boolean hasPermission(Object permission) {
            return true;
        }

        @Override
        public URL getResource(String name) {
            return getClassLoader().getResource(name);
        }

        @Override
        public Class<?> loadClass(String name) throws ClassNotFoundException {
            return getClassLoader().loadClass(name);
        }

        @Override
        public Enumeration<URL> getResources(String name) throws IOException {
            return getClassLoader().getResources(name);
        }

        @Override
        public Enumeration<String> getEntryPaths(String path) {
            throw new UnsupportedOperationException();
        }

        @Override
        public URL getEntry(String path) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Enumeration<URL> findEntries(String path, String filePattern, boolean recurse) {
            throw new UnsupportedOperationException();
        }

        @Override
        public long getLastModified() {
            return 1L;
        }

        @Override
        public BundleContext getBundleContext() {
            throw new UnsupportedOperationException();
        }

        @Override
        public Map<X509Certificate, List<X509Certificate>> getSignerCertificates(int signersType) {
            return Collections.emptyMap();
        }

        @Override
        @SuppressWarnings("unchecked")
        public <T> T adapt(Class<T> clazz) {
            if (clazz.equals(BundleRevision.class)) {
                return (T) new BundleRevisionImpl();
            } else if (clazz.equals(BundleWiring.class)) {
                return (T) new BundleWiringImpl();
            } else {
                return null;
            }
        }

        @Override
        public File getDataFile(String filename) {
            return null;
        }

        @Override
        public int compareTo(Bundle o) {
            return Long.compare(getBundleId(), o.getBundleId());
        }
    }

    private class BundleRevisionImpl implements BundleRevision {
        @Override
        public String getSymbolicName() {
            return this.getClass().getName();
        }

        @Override
        public List<BundleRequirement> getDeclaredRequirements(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Version getVersion() {
            return Version.emptyVersion;
        }

        @Override
        public BundleWiring getWiring() {
            return bundleWiringImpl;
        }

        @Override
        public List<BundleCapability> getDeclaredCapabilities(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public int getTypes() {
            return 0;
        }

        @Override
        public Bundle getBundle() {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<Capability> getCapabilities(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<Requirement> getRequirements(String p1) {
            throw new UnsupportedOperationException();
        }
    }

    private class BundleWiringImpl implements BundleWiring {
        @Override
        public List<URL> findEntries(String p1, String p2, int p3) {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<Wire> getRequiredResourceWires(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<Capability> getResourceCapabilities(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public boolean isCurrent() {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<BundleWire> getRequiredWires(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<BundleCapability> getCapabilities(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<Wire> getProvidedResourceWires(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<BundleWire> getProvidedWires(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public BundleRevision getRevision() {
            throw new UnsupportedOperationException();
        }

        @Override
        public List<Requirement> getResourceRequirements(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public boolean isInUse() {
            throw new UnsupportedOperationException();
        }

        @Override
        public Collection<String> listResources(String p1, String p2, int p3) {
            throw new UnsupportedOperationException();
        }

        @Override
        public ClassLoader getClassLoader() {
            return ClassLoaderOsgiFramework.this.getClassLoader();
        }

        @Override
        public List<BundleRequirement> getRequirements(String p1) {
            throw new UnsupportedOperationException();
        }

        @Override
        public BundleRevision getResource() {
            throw new UnsupportedOperationException();
        }

        @Override
        public Bundle getBundle() {
            throw new UnsupportedOperationException();
        }
    }

    private class SystemBundleImpl extends BundleImpl {
        @Override
        public long getBundleId() {
            return 0L;
        }

        @Override
        public Version getVersion() {
            return Version.emptyVersion;
        }

        @Override
        public Dictionary<String, String> getHeaders() {
            Hashtable<String, String> ret = new Hashtable<>();
            ret.put(OsgiHeader.APPLICATION, StandaloneContainerApplication.class.getName());
            return ret;
        }
    }

    private class JarBundleImpl extends BundleImpl {
        private final long bundleId;
        private final Dictionary<String, String> headers;

        JarBundleImpl(URL location) throws IOException {
            this.bundleId = (long) nextBundleId.getAndIncrement();
            this.headers = retrieveHeaders(location);
        }

        @Override
        public long getBundleId() {
            return bundleId;
        }

        @Override
        public Dictionary<String, String> getHeaders() {
            return headers;
        }

        @Override
        public String getSymbolicName() {
            return headers.get("Bundle-SymbolicName");
        }

        @Override
        public Version getVersion() {
            return Version.parseVersion(headers.get("Bundle-Version"));
        }

        private Dictionary<String, String> retrieveHeaders(URL location) throws IOException {
            try (JarFile jarFile = new JarFile(location.getFile())) {
                Attributes attributes = jarFile.getManifest().getMainAttributes();
                Hashtable<String, String> ret = new Hashtable<>();
                attributes.forEach((k, v) -> ret.put(k.toString(), v.toString()));
                return ret;
            }
        }
    }

    private class BundleContextImpl implements BundleContext {
        @Override
        public String getProperty(String key) {
            return null;
        }

        @Override
        public Bundle getBundle() {
            return systemBundleImpl;
        }

        @Override
        public Bundle installBundle(String location, InputStream input) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Bundle installBundle(String location) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Bundle getBundle(long id) {
            return systemBundleImpl;
        }

        @Override
        public Bundle[] getBundles() {
            return new Bundle[] { systemBundleImpl };
        }

        @Override
        public Bundle getBundle(String location) {
            return systemBundleImpl;
        }

        @Override
        public void addServiceListener(ServiceListener listener, String filter) {
        }

        @Override
        public void addServiceListener(ServiceListener listener) {
        }

        @Override
        public void removeServiceListener(ServiceListener listener) {
        }

        @Override
        public void addBundleListener(BundleListener listener) {
        }

        @Override
        public void removeBundleListener(BundleListener listener) {
        }

        @Override
        public void addFrameworkListener(FrameworkListener listener) {
        }

        @Override
        public void removeFrameworkListener(FrameworkListener listener) {
        }

        @Override
        public ServiceRegistration<?> registerService(String[] classes, Object service, Dictionary<String, ?> properties) {
            throw new UnsupportedOperationException();
        }

        @Override
        public ServiceRegistration<?> registerService(String clazz, Object service, Dictionary<String, ?> properties) {
            return null;
        }

        @Override
        public <S> ServiceRegistration<S> registerService(Class<S> clazz, S service, Dictionary<String, ?> properties) {
            throw new UnsupportedOperationException();
        }

        @Override
        public ServiceReference<?>[] getServiceReferences(String clazz, String filter) {
            throw new UnsupportedOperationException();
        }

        @Override
        public ServiceReference<?>[] getAllServiceReferences(String clazz, String filter) {
            throw new UnsupportedOperationException();
        }

        @Override
        public ServiceReference<?> getServiceReference(String clazz) {
            throw new UnsupportedOperationException();
        }

        @Override
        public <S> ServiceReference<S> getServiceReference(Class<S> clazz) {
            throw new UnsupportedOperationException();
        }

        @Override
        public <S> Collection<ServiceReference<S>> getServiceReferences(Class<S> clazz, String filter) {
            return new ArrayList<>();
        }

        @Override
        public <S> S getService(ServiceReference<S> reference) {
            throw new UnsupportedOperationException();
        }

        @Override
        public boolean ungetService(ServiceReference<?> reference) {
            throw new UnsupportedOperationException();
        }

        @Override
        public File getDataFile(String filename) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Filter createFilter(String filter) {
            throw new UnsupportedOperationException();
        }

        @Override
        public <S> ServiceRegistration<S> registerService(Class<S> aClass, ServiceFactory<S> serviceFactory,
                Dictionary<String, ?> dictionary) {
            throw new UnsupportedOperationException();
        }

        @Override
        public <S> ServiceObjects<S> getServiceObjects(ServiceReference<S> serviceReference) {
            throw new UnsupportedOperationException();
        }
    }
}
